test/ffmpeg-qsv/transcode/transcoder.py

```diff
@@ -120,15 +137,25 @@ def gen_output_opts(self):
     for n, output in enumerate(self.outputs):
       codec = output["codec"]
       mode = output["mode"]
-
+      width = output.get("width", 0)
+      height = output.get("height", 0)
       _, _, ffparms = self.get_requirements_data("encode", codec, mode)
       assert ffparms is not None, "failed to find a suitable encoder"
 
       ffcodec, ext = ffparms
 
       for channel in xrange(output.get("channels", 1)):
         if "hw" == mode and "sw" == self.mode:
-          opts += " -vf hwupload=extra_hw_frames=64,format=qsv"
+          if width > 0 and height > 0:
+            opts += " -vf 'hwupload=extra_hw_frames=64,format=qsv,vpp_qsv=w={}:h={}'".format(width, height)
+          else:
+            opts += " -vf hwupload=extra_hw_frames=64,format=qsv"
+        elif "sw" == mode and "hw" == self.mode:
+          if width > 0 and height > 0:
+            opts += " -vf 'hwdownload,format=nv12,scale=w={}:h={}'".format(width, height)
+        else:
+          if width > 0 and height > 0:
+            opts += " -vf 'hwupload=extra_hw_frames=64,format=qsv,vpp_qsv=w={}:h={}'".format(width, height)
```

Comment:
There are several common parts to each mode combination and can be de-coupled from scaling.  Seems to me that this can be handled better without the chaos of nested if/else conditions.